synedrion/src/cggmp21/key_refresh.rs

```diff
@@ -1,321 +1,632 @@
 
-use alloc::{
-    collections::{BTreeMap, BTreeSet},
-    format,
-    string::String,
-    vec::Vec,
+use alloc::collections::{BTreeMap, BTreeSet};
+use core::{
+    fmt::{self, Debug, Display},
+    marker::PhantomData,
 };
-use core::{fmt::Debug, marker::PhantomData};
 
 use crypto_bigint::BitOps;
 use manul::protocol::{
     Artifact, BoxedRound, Deserializer, DirectMessage, EchoBroadcast, EntryPoint, FinalizeOutcome, LocalError,
-    NormalBroadcast, PartyId, Payload, Protocol, ProtocolError, ProtocolMessagePart, ProtocolValidationError,
-    ReceiveError, Round, RoundId, Serializer,
+    MessageValidationError, NormalBroadcast, PartyId, Payload, Protocol, ProtocolError, ProtocolMessage,
+    ProtocolMessagePart, ProtocolValidationError, ReceiveError, RequiredMessageParts, RequiredMessages, Round, RoundId,
+    Serializer,
 };
 use rand_core::CryptoRngCore;
 use serde::{Deserialize, Serialize};
 
 use super::{
-    conversion::{secret_scalar_from_uint, secret_unsigned_from_scalar},
-    entities::{AuxInfo, KeyShareChange, PublicAuxInfo, SecretAuxInfo},
+    entities::{AuxInfo, KeyShareChange, PublicAuxInfo, PublicAuxInfos, SecretAuxInfo},
     params::SchemeParams,
     sigma::{FacProof, ModProof, PrmProof, SchCommitment, SchProof, SchSecret},
 };
 use crate::{
     curve::{secret_split, Point, Scalar},
     paillier::{
-        Ciphertext, CiphertextWire, PaillierParams, PublicKeyPaillier, PublicKeyPaillierWire, RPParams, RPParamsWire,
-        RPSecret, SecretKeyPaillier, SecretKeyPaillierWire,
+        PaillierParams, PublicKeyPaillier, PublicKeyPaillierWire, RPParams, RPParamsWire, RPSecret, SecretKeyPaillier,
+        SecretKeyPaillierWire,
     },
     tools::{
         bitvec::BitVec,
-        hashing::{Chain, FofHasher, HashOutput},
-        DowncastMap, Secret, Without,
+        hashing::{Chain, FofHasher, HashOutput, XofHasher},
+        protocol_shortcuts::{verify_that, DeserializeAll, DowncastMap, GetRound, MapValues, SafeGet, Without},
+        Secret,
     },
 };
 
 /// A protocol for generating auxiliary information for signing,
 /// and a simultaneous generation of updates for the secret key shares.
 #[derive(Debug)]
-pub struct KeyRefreshProtocol<P: SchemeParams, I: PartyId>(PhantomData<(P, I)>);
+pub struct KeyRefreshProtocol<P: SchemeParams, Id: PartyId>(PhantomData<(P, Id)>);
 
-impl<P: SchemeParams, I: PartyId> Protocol for KeyRefreshProtocol<P, I> {
-    type Result = (KeyShareChange<P, I>, AuxInfo<P, I>);
-    type ProtocolError = KeyRefreshError<P>;
+impl<P: SchemeParams, Id: PartyId> Protocol<Id> for KeyRefreshProtocol<P, Id> {
+    type Result = (KeyShareChange<P, Id>, AuxInfo<P, Id>);
+    type ProtocolError = KeyRefreshError<P, Id>;
+
+    fn verify_direct_message_is_invalid(
+        deserializer: &Deserializer,
+        round_id: &RoundId,
+        message: &DirectMessage,
+    ) -> Result<(), MessageValidationError> {
+        match round_id {
+            r if r == &1 => message.verify_is_some(),
```

Comment:
Could be `r if r == &1 || r == &2 => message.verify_is_some(),` I guess.

Reply:
I wanted to keep the visual structure of one round per line, even if the checks repeat.